Introduce JPQL/HQL parser for JPQ queries

gregturn opened the pull request:
Tickets that require attention once this is done:

- [ ] https://github.com/spring-projects/spring-data-jpa/issues/2831

mp911de commented on spring-data-jpa/src/main/java/org/springframework/data/jpa/repository/query/QueryParser.java:
```diff
@@ -0,0 +1,159 @@
+	 */
+	abstract boolean hasConstructor(ParserRuleContext parsedQuery);
+
+	private static final Pattern PUNCTUATION_PATTERN = Pattern.compile(".*((?![._])[\\p{Punct}|\\s])");
```
Nit: These are at a strange position, should be on top of the class.

gregturn replied:
I can move this toward the top.

mp911de commented on spring-data-jpa/src/main/java/org/springframework/data/jpa/repository/query/QueryParsingEnhancer.java:
```diff
@@ -0,0 +1,208 @@
+
+	private static final Log LOG = LogFactory.getLog(QueryParsingEnhancer.class);
+
+	private final QueryParser queryParser;
```
This is a neat approach reusing existing abstractions 👍

mp911de commented on spring-data-jpa/src/main/java/org/springframework/data/jpa/repository/query/QueryParsingEnhancer.java:
```diff
@@ -0,0 +1,208 @@
+
+			return queryParser.createQuery(parsedQuery, sort);
+		} catch (QueryParsingSyntaxError e) {
+			LOG.warn(e);
```
Logging and throwing an exception is duplicate exception handling.

mp911de commented on spring-data-jpa/src/main/java/org/springframework/data/jpa/repository/query/QueryParsingEnhancer.java:
```diff
@@ -0,0 +1,208 @@
+
+			return queryParser.findAlias(parsedQuery);
+		} catch (QueryParsingSyntaxError e) {
+			LOG.warn(e);
```
Same here.

mp911de commented on spring-data-jpa/src/main/java/org/springframework/data/jpa/repository/query/QueryParsingEnhancer.java:
```diff
@@ -0,0 +1,208 @@
+	 * @return empty set
+	 */
+	@Override
+	public Set<String> getJoinAliases() {
```
`getJoinAliases` seems to be not used in production code. We should deprecate it and make it a `default` method in the interface.

mp911de commented on spring-data-jpa/src/main/java/org/springframework/data/jpa/repository/query/QueryParsingSyntaxError.java:
```diff
@@ -0,0 +1,41 @@
+ */
+class QueryParsingSyntaxError extends ParseCancellationException {
+
+	public QueryParsingSyntaxError() {}
```
Can we attach the original SQL query to the exception for more context? Also, this exception should be public so that users can handle this exception.

gregturn replied:
We can definitely attach the original query to this exception.

mp911de commented on spring-data-jpa/src/main/java/org/springframework/data/jpa/repository/query/QueryParsingSyntaxError.java:
```diff
@@ -0,0 +1,41 @@
+ * @author Greg Turnquist
+ * @since 3.1
+ */
+class QueryParsingSyntaxError extends ParseCancellationException {
```
I'm inclined to let the exception extend `InvalidDataAccessResourceUsageException` (similar to `BadSqlGrammarException`) instead of an antlr type.

gregturn replied:
I'm not sure. `ParseCancellationException` is a signal to ANTLR to stop parsing. I can try it out and see what happens.

mp911de commented on spring-data-jpa/src/main/java/org/springframework/data/jpa/repository/query/QueryParsingSyntaxErrorListener.java:
```diff
@@ -0,0 +1,35 @@
+ * @author Greg Turnquist
+ * @since 3.1
+ */
+class QueryParsingSyntaxErrorListener extends BaseErrorListener {
```
We could turn this into an enum or create a singleton to avoid instance creation on each usage.

gregturn replied:
We could, unless we want to pass in the original query. This then becomes a factory operation. I can poke at it to see what's possible.

mp911de commented on spring-data-jpa/src/main/java/org/springframework/data/jpa/repository/query/StringQuery.java:
```diff
@@ -68,6 +67,7 @@ class StringQuery implements DeclaredQuery {
 	 *
 	 * @param query must not be {@literal null} or empty.
 	 */
+	@Deprecated
```
Why is this deprecated? This leaves us without a deprecated constructor.

gregturn replied:
I think this is relic from when I tried to introduce "modes". I'll remove it.

mp911de commented on spring-data-jpa/src/main/java/org/springframework/data/jpa/repository/query/JpqlQueryTransformer.java:
```diff
@@ -0,0 +1,2437 @@
+ * @author Greg Turnquist
+ * @since 3.1
+ */
+class JpqlQueryTransformer extends JpqlBaseVisitor<List<QueryParsingToken>> {
```
This and the HQL transformer require additional work.

1. The many occurrences of `getText()` (e.g. `new QueryParsingToken(ctx.FROM().getText(), true)`) indicate that the code isn't ideal. We're creating a lot of objects for mostly static components that we could cache. Also, regular `getText()` calls could be abstracted away to avoid duplications.
2. This class combines query parsing/rendering along with our domain-specific tweaks to introduce sorting, count query projections. Ideally, we can encapsulate all base-work that is necessary for query rendering into a base transformer class and have transformers for specific tasks (count query derivation, sorting, …). To put things into perspective: `QueryParser` has individual methods for each use case, while this class combines all use-cases.
3. HQL and JPQL share a lot of common ground. It isn't ideal that we have so many duplications. Also, when comparing both files side-by-side, it's impossible to put both files next to each other to determine differences. Going forward, this will be the method to detect, if we apply a change to one, that we need to apply the same change to the other transformer as well. We need to fix that. First step would be to align method orders (e.g. that the HQL `visitJoinType` method is at the same place where the JPQL `visitJoin_spec` method is). While this is purely a style aspect from the technical perspective, it is something that decides over how difficult it becomes to maintain parsers.

mp911de reviewed (changes requested):
I like where this PR is heading. I left a few initial comments about the structure and design of the parsers. Once these addressed, we can dig into more detail of parsing and rendering queries to make more progress.

How do we ensure that our visitors make sure of all permutations and objects (i.e. that we haven't forgotten to render out a particular case that is possible with JPQL/HQL but we don't have such a case in our tests)?

gregturn commented on spring-data-jpa/src/main/java/org/springframework/data/jpa/repository/query/QueryParser.java:
```diff
@@ -0,0 +1,159 @@
+ * @author Greg Turnquist
+ * @since 3.1
+ */
+abstract class QueryParser {
```
I like that.

gregturn commented on spring-data-jpa/src/main/java/org/springframework/data/jpa/repository/query/QueryParser.java:
```diff
@@ -0,0 +1,159 @@
+
+	private final DeclaredQuery declaredQuery;
+
+	QueryParser(DeclaredQuery declaredQuery) {
```
Agreed.

gregturn commented on spring-data-jpa/src/main/java/org/springframework/data/jpa/repository/query/QueryParser.java:
```diff
@@ -0,0 +1,159 @@
+	 * @param parsedQuery
+	 * @param sort can be {@literal null}
+	 */
+	String createQuery(ParserRuleContext parsedQuery, Sort sort) {
```
The is because invoking the ANTLR-generate `visit()` method requires passing in one of the generated context objects. ALL of their context classes implement `ParserRuleContext`, so this was my way to make HQL and JPQL both have a common API.

Basically:

```java
HqlLexer lexer = new HqlLexer(CharStreams.fromString(query));
HqlParser parser = new HqlParser(new CommonTokenStream(lexer));

parser.addErrorListener(new QueryParsingSyntaxErrorListener());

return parser.start();
```
...is the ANTLR pattern to parse a query and get handed back an AST.

```java
new HqlQueryTransformer(sort).visit(parsedQuery)
```
...is the ANTLR way to then apply one of their visitors to that AST. I'm trying to encapsulate this flow inside of `QueryParser`.

The reasons I have this "visible" is so that `QueryEnhancer` can apply various strategies based on what function it's doing. For instance, when you do `applySorting`, it's supposed to raise an `IllegalArgumentException`. 

```java
@Override
public String applySorting(Sort sort) {

	try {
		ParserRuleContext parsedQuery = queryParser.parse();

		if (parsedQuery == null) {
			return "";
		}

		return queryParser.createQuery(parsedQuery, sort);
	} catch (QueryParsingSyntaxError e) {
		LOG.warn(e);
		throw new IllegalArgumentException(e);
	}
}
```

However, if you run `detectAlias`, it's does NOT throw anything, but instead will simply return `null` for an invalid query.

```java
@Override
public String detectAlias() {

	try {
		ParserRuleContext parsedQuery = queryParser.parse();

		if (parsedQuery == null) {

			LOG.warn("Failed to parse " + queryParser.getQuery() + ". See console for more details.");
			return null;
		}

		return queryParser.findAlias(parsedQuery);
	} catch (QueryParsingSyntaxError e) {
		LOG.warn(e);
		return null;
	}
}
```

This is where it would be nice if we had a chance to refactor `QueryEnhancer`. But that feels like too big a change. I figured if we kept all the ANTLR specifics inside `QueryParsingEnhancer`, we could minimize that leakage.

gregturn commented:
@mp911de I've processed all feedback items discussed this morning. For lack of a better name, I have split `HqlQueryTransformer` into `HqlQueryRenderer` and `HqlQueryTransformer`. All of the HQL specification tests I wrote earlier I now have running in a separate test class (`HqlQueryRendererTests`) to ensure that it renders queries exactly as they came in. 

This has made `HqlQueryTransformer` much lighterweight, and eases our ability to make overrides and adjustments. And I implemented the counterpart to this for JPQL-based queries.

I also moved all the ANTLR  specifics into `JpaQueryParser` such that the `QueryParsingEnhancer` doesn't have to see it. And the internal ANTLR-based methods are all now `protected`, further sealing this off.

I haven't gotten (yet) to any caching solution, but I have verified everything else is functional.

mp911de commented:
I added a polishing commit to encapsulate components even more and reuse parser results as the initial parser yielded 1700 parse operations per second while the regex-one did 120000 ops/sec.
